### ISSUE TITLE
Add support for SuppressMessageAttribute.MessageId 

### ISSUE THREAD

**Issue by JiriCepelkaFirstLineSoftware** [dotnet/roslyn#38827]
There is no option to avoid consequent violation to suppressed rule applied to method. To do so analyzers should support [_MessageId_](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.suppressmessageattribute.messageid?view=netframework-4.8#System_Diagnostics_CodeAnalysis_SuppressMessageAttribute_MessageId) property.

    [SuppressMessage("SomeCat", "SomeRule")]
    void SomeMethod()
    {
     // line violating SomeRule
    } 

Relevant articles:
1. SuppressMessageAttribute Class/[Examples](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.suppressmessageattribute?view=netframework-4.8#examples)
2. Suppress code analysis warnings/[Suppress selective violations within a method body](https://docs.microsoft.com/en-us/visualstudio/code-quality/in-source-suppression-overview?view=vs-2019#suppress-selective-violations-within-a-method-body)

Further more if _MessageId_ is specified, _SuppressMessageAttribute_ is ignored, i.e. error is reported anyway (by _Managed Binary Analysis_).

`[SuppressMessage("SomeCat", "SomeRule, MessageId = "someVar")]`

_Pragma warning disable_ is not symbol based. Since using statements denotes code blocks, pragma can disable consequent violations also.

**Comment by mavasani** [comment]
This is by design - see https://github.com/dotnet/roslyn-analyzers/issues/2865. You should use pragma warning suppressions for suppressing Roslyn analyzer diagnostics for specific lines.

@JiriCepelkaFirstLineSoftware You mentioned that you are unable to suppress CA2202 with pragma directives, but this rule has been Won't fixed: https://github.com/dotnet/roslyn-analyzers/issues/528. It seems you are running [legacy FxCop](https://docs.microsoft.com/en-us/visualstudio/code-quality/static-code-analysis-for-managed-code-overview?view=vs-2019) version to get your analysis results, which has been deprecated: https://docs.microsoft.com/en-us/visualstudio/code-quality/fxcop-analyzers?view=vs-2019

**Comment by JiriCepelkaFirstLineSoftware** [comment]
I prepared test [projects](uzivatel919.sweb.cz/SuppressMessageAttributeAndPragma.rar). I am using VS 2017C 15.9.16. Project is .NET Framework 4.7 console app with MS all rules.

1) When I use default analyzer pragma is ignored. 
2) With installed Microsoft.CodeAnalysis.FxCopAnalyzers 2.9.4 pragma is ignored.

So pragma works only in VS 2019?

**Comment by mavasani** [comment]
How are you getting CA2202 diagnostics, given CA2202 is not reported by Microsoft.CodeAnalysis.FxCopAnalyzers? Are you explicitly executing "Run code analysis", which runs legacy code analysis?

**Comment by JiriCepelkaFirstLineSoftware** [comment]
I do Analyze/Run Code Analysis/On Solution. I feel this a natural way of analysing the solution using VS.

**Comment by mavasani** [comment]
@JiriCepelkaFirstLineSoftware Please go through the FAQs here: https://docs.microsoft.com/en-us/visualstudio/code-quality/fxcop-analyzers-faq?view=vs-2019

Note that "Run Code Analysis" command will be removed for Managed projects very soon as it always executes legacy code analysis and is on path to deprecation. Source based Roslyn Analyzers run by default when you type in VS or build your project. You can configure this behavior in Code Analysis properties page:

![image](https://user-images.githubusercontent.com/10605811/65602023-05dd0e00-df58-11e9-8bb3-ccbd5621af9e.png)

**Comment by JiriCepelkaFirstLineSoftware** [comment]
Screenshot you provided comes form VS2019. What is the future for VS2017?

**Comment by mavasani** [comment]
> Screenshot you provided comes form VS2019. What is the future for VS2017?

Roslyn analyzers work completely fine even on VS2017. You should avoid executing "Run Code Analysis" and instead just install the analyzer package. The above 2 checkboxes (and functionality to conditionally turn off analyzers in live analysis or build) are driven by very simple MSBuild target in VS2019:

```
<PropertyGroup>
    <IsCSharpOrVbProject Condition="'$(Language)'=='C#' or '$(Language)'=='VB'">true</IsCSharpOrVbProject>
    <RunAnalyzersDuringLiveAnalysis Condition="'$(RunAnalyzersDuringLiveAnalysis)' == ''">$(IsCSharpOrVbProject)</RunAnalyzersDuringLiveAnalysis>
    <RunAnalyzersDuringBuild Condition="'$(RunAnalyzersDuringBuild)' == ''">$(IsCSharpOrVbProject)</RunAnalyzersDuringBuild>
  </PropertyGroup>

  <Target Name="ComputeRunAnalyzers"
           BeforeTargets="CoreCompile"
           Condition="'$(RunAnalyzers)' == ''">
    <PropertyGroup>
      <IsAnyDesignTimeBuild Condition="'$(DesignTimeBuild)' == 'true' or '$(BuildingProject)' != 'true'">true</IsAnyDesignTimeBuild>
      <RunAnalyzers Condition="'$(IsAnyDesignTimeBuild)' == 'true' and '$(RunAnalyzersDuringLiveAnalysis)' == 'false'">false</RunAnalyzers>
      <RunAnalyzers Condition="'$(IsAnyDesignTimeBuild)' != 'true' and '$(RunAnalyzersDuringBuild)' == 'false'">false</RunAnalyzers>
    </PropertyGroup>
  </Target>
  
  <Target Name="DisableAnalyzers"
           AfterTargets="ComputeRunAnalyzers"
           Condition="'$(RunAnalyzers)' == 'false'">
    <!-- 
       Disable analyzers via an MSBuild property settable on the command line. 
     -->
    <ItemGroup>
      <Analyzer Remove="@(Analyzer)" />
    </ItemGroup>
  </Target>
```

You should be able to add the above to your root targets files and then configure `RunAnalyzersDuringLiveAnalysis` or `RunAnalyzersDuringBuild` in individual project files to get the same functionality.

**Comment by JiriCepelkaFirstLineSoftware** [comment]
Ah, that’s fine. I cannot verify today but tomorrow I’ll let you know the results.

**Comment by mavasani** [comment]
Closing this issue as by design - please let us know if you need any other assistance. Thanks!

**Comment by JiriCepelkaFirstLineSoftware** [comment]
I’m failing in getting the analyzers run. [Project](http://uzivatel919.sweb.cz/SuppressMessageAttributeAndPragma%20%E2%80%93%20analyzers.rar). [Build.props](http://uzivatel919.sweb.cz/Directory.Build.props).

**Comment by JiriCepelkaFirstLineSoftware** [comment]
@mavasani,
I set up [analyzers](http://uzivatel919.sweb.cz/SuppressMessageAttributeAndPragma%20%E2%80%93%20analyzers.rar) with _Directory.Build.props_ as you told me but _pragma disable warning_ is still ignored.

Can you verify? (Perhaps you will need to change something in solution to force analyzers to run on build.)

**Comment by mavasani** [comment]
@JiriCepelkaFirstLineSoftware CA2202 is not reported by FxCop analyzers package. So if you are still getting CA2202 diagnostic, you are executing legacy FxCop analyzers. Can you please remove the below property from your project file:

`<RunCodeAnalysis>true</RunCodeAnalysis>`

This property forces legacy code analysis to execute. See https://docs.microsoft.com/en-us/visualstudio/code-quality/fxcop-analyzers-faq?view=vs-2019#does-the-runcodeanalysis-msbuild-project-property-run-analyzers

**Comment by JiriCepelkaFirstLineSoftware** [comment]
The element is set up 

`<RunCodeAnalysis>false</RunCodeAnalysis>`.

Nonetheless you’re right. Error originates from [_Managed Binary Analysis](http://uzivatel919.sweb.cz/Analyzers.png)_. But none of the [analyzers](http://uzivatel919.sweb.cz/CA2202.png) from package _Microsoft.CodeAnalysis.FxCopAnalyzers 2.9.4_ has this rule.

**Comment by JiriCepelkaFirstLineSoftware** [comment]
@mavasani,

I found an issue [Port FxCop rule CA2202: DoNotDisposeObjectsMultipleTimes](https://github.com/dotnet/roslyn-analyzers/issues/528). So _FxCop Analyzers_ don’t provide such validation.

On issue you’re [noting](https://github.com/dotnet/roslyn-analyzers/issues/528#issuecomment-451944586-permalink) some sensitive analysis. In _Managed Binary Analysis_ it is not so sensitive – reports false positives.

The rules is discontinued. What is your recommendation? Is still this issue out of place?
